RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+      maxEtaForTOB_(iConfig.getParameter<double>("maxEtaForTOB")),
+      updator_(new KFUpdator()),
+      measurementTrackerTag_(
+          consumes<MeasurementTrackerEvent>(iConfig.getParameter<edm::InputTag>("MeasurementTrackerEvent"))),
+      theCategory_(std::string("Muon|RecoMuon|TSGForOIDNN")),
+      maxHitlessSeedsIP_(iConfig.getParameter<uint32_t>("maxHitlessSeedsIP")),
+      maxHitlessSeedsMuS_(iConfig.getParameter<uint32_t>("maxHitlessSeedsMuS")),
+      maxHitDoubletSeeds_(iConfig.getParameter<uint32_t>("maxHitDoubletSeeds")),
+      getStrategyFromDNN_(iConfig.getParameter<bool>("getStrategyFromDNN")),
+      useRegressor_(iConfig.getParameter<bool>("useRegressor")),
+      etaSplitForDnn_(iConfig.getParameter<double>("etaSplitForDnn")),
+      dnnMetadataPath_(iConfig.getParameter<std::string>("dnnMetadataPath")) {
+  if (getStrategyFromDNN_) {
+    edm::FileInPath dnnMetadataPath(dnnMetadataPath_);
+    pt::read_json(dnnMetadataPath.fullPath(), metadata);
+    tensorflow::setLogging("2");
+
+    if (useRegressor_) {
+      // use regressor
+      dnnModelPath_HB_ = metadata.get<std::string>("HB.dnnmodel_path");
+      edm::FileInPath dnnPath_HB(dnnModelPath_HB_);
+      graphDef_HB_ = tensorflow::loadGraphDef(dnnPath_HB.fullPath());
+      tf_session_HB_ = tensorflow::createSession(graphDef_HB_);
+
+      dnnModelPath_HLIP_ = metadata.get<std::string>("HLIP.dnnmodel_path");
+      edm::FileInPath dnnPath_HLIP(dnnModelPath_HLIP_);
+      graphDef_HLIP_ = tensorflow::loadGraphDef(dnnPath_HLIP.fullPath());
+      tf_session_HLIP_ = tensorflow::createSession(graphDef_HLIP_);
+
+      dnnModelPath_HLMuS_ = metadata.get<std::string>("HLMuS.dnnmodel_path");
+      edm::FileInPath dnnPath_HLMuS(dnnModelPath_HLMuS_);
+      graphDef_HLMuS_ = tensorflow::loadGraphDef(dnnPath_HLMuS.fullPath());
+      tf_session_HLMuS_ = tensorflow::createSession(graphDef_HLMuS_);
+    } else {
+      // use classifier (default)
+      dnnModelPath_barrel_ = metadata.get<std::string>("barrel.dnnmodel_path");
+      edm::FileInPath dnnPath_barrel(dnnModelPath_barrel_);
+      graphDef_barrel_ = tensorflow::loadGraphDef(dnnPath_barrel.fullPath());
+      tf_session_barrel_ = tensorflow::createSession(graphDef_barrel_);
+
+      dnnModelPath_endcap_ = metadata.get<std::string>("endcap.dnnmodel_path");
+      edm::FileInPath dnnPath_endcap(dnnModelPath_endcap_);
+      graphDef_endcap_ = tensorflow::loadGraphDef(dnnPath_endcap.fullPath());
+      tf_session_endcap_ = tensorflow::createSession(graphDef_endcap_);
+    }
+  }
+  produces<std::vector<TrajectorySeed> >();
+}
+
+TSGForOIDNN::~TSGForOIDNN() {
+  if (getStrategyFromDNN_) {
+    if (useRegressor_) {
+      tensorflow::closeSession(tf_session_HB_);
+      tensorflow::closeSession(tf_session_HLIP_);
+      tensorflow::closeSession(tf_session_HLMuS_);
+      delete graphDef_HB_;
+      delete graphDef_HLIP_;
+      delete graphDef_HLMuS_;
```

Comment:
to be removed after these are made to `unique_ptr`
```suggestion
```

Reply:
Done!